### PR TITLE
[BEAM-10812] fix(beam/sdks/go): fix buffer limit on textio

### PR DESCRIPTION
**Pull request by ceocoder**
bufio.Scanner is limited to 4kb by default, for arbitary line length it
is recommended to use bufio.Reader and split by '\n'

Ref: https://github.com/golang/go/blob/master/src/bufio/scan.go#L27

issue identifyed by @salrashid123

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace
--- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
Build python source distribution and wheels
Python tests
Java tests

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by ceocoder** [comment]
R: @youngoli
cc: @TheNeuralBit

File: sdks/go/pkg/beam/io/textio/textio.go [hunk 1]
```diff
@@ -93,11 +96,18 @@ func readFn(ctx context.Context, filename string, emit func(string)) error {
 	}
 	defer fd.Close()
 
-	scanner := bufio.NewScanner(fd)
-	for scanner.Scan() {
-		emit(scanner.Text())
+	rd := bufio.NewReader(fd)
+	for {
+		line, err := rd.ReadString('\n')
+		if err == io.EOF {
```
Comment (youngoli) [inline sdks/go/pkg/beam/io/textio/textio.go:102]: The documentation for ReadString states:

> If ReadString encounters an error before finding a delimiter, it returns the data read before the error and the error itself (often io.EOF)

So it sounds like you'd want to emit the line here once before breaking, maybe with a quick check `if len(line) != 0`.

File: sdks/go/pkg/beam/io/textio/textio.go [hunk 2]
```diff
@@ -93,11 +96,18 @@ func readFn(ctx context.Context, filename string, emit func(string)) error {
 	}
 	defer fd.Close()
 
-	scanner := bufio.NewScanner(fd)
-	for scanner.Scan() {
-		emit(scanner.Text())
+	rd := bufio.NewReader(fd)
+	for {
+		line, err := rd.ReadString('\n')
```
Comment (youngoli) [inline sdks/go/pkg/beam/io/textio/textio.go:101]: One thing that caught my eye on the ReadString description was that the string returned includes the delimiter, which differs from the old scanner implementation which strips the delimiter. To keep parity between the old and new behavior you'll probably want to strip the newline before emitting the string.

File: sdks/go/pkg/beam/io/textio/textio_test.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+		receivedLines = append(receivedLines, line)
+	}
+
+	err := readFn(nil, f, getLines)
+	if err != nil {
+		t.Fatalf("failed with %v", err)
+	}
+	if len(receivedLines) != 1 {
+		t.Fatalf("received %v lines", len(receivedLines))
+	}
```
Comment (youngoli) [inline sdks/go/pkg/beam/io/textio/textio_test.go:39]: You should mention what the expected number of lines is, for example:
```suggestion
	want := 1
	if len(receivedLines) != want {
		t.Fatalf("received %v lines, wanted %v", len(receivedLines), want)
	}
```

File: sdks/go/pkg/beam/io/textio/testfile.txt [hunk 4]
```diff
@@ -0,0 +1 @@
...ooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdffooooasdfasdfasdfasdfasdfdsfdfooooasdfasdfasdfasdfasdfdsfdfff
```
Comment (lostluck) [inline sdks/go/pkg/beam/io/textio/testfile.txt:1]: We have a license check (the RAT precommit), which requires that all files have the Apache licence header, which this file is triggering.

To have the check ignore this new test data file, please add 
    "**/go/**/*.txt",
to
https://github.com/apache/beam/blob/master/build.gradle#L81

Or move this test data to .../sdks/go/data/...

File: sdks/go/pkg/beam/io/textio/textio_test.go (hunk 3 above)
Comment (ceocoder) [reply sdks/go/pkg/beam/io/textio/textio_test.go]: done

File: sdks/go/pkg/beam/io/textio/testfile.txt (hunk 4 above)
Comment (ceocoder) [reply sdks/go/pkg/beam/io/textio/testfile.txt]: done. moved to `go/data/..`

File: sdks/go/pkg/beam/io/textio/textio.go [hunk 5]
```diff
@@ -79,6 +81,7 @@ func expandFn(ctx context.Context, glob string, emit func(string)) error {
 }
 
 func readFn(ctx context.Context, filename string, emit func(string)) error {
+	fmt.Println(filename)
```
[the remark this answers is not in the document]
Comment (ceocoder) [reply sdks/go/pkg/beam/io/textio/textio.go]: yikes.  yes - this was my bad.

File: sdks/go/pkg/beam/io/textio/textio.go (hunk 1 above)
Comment (ceocoder) [reply sdks/go/pkg/beam/io/textio/textio.go:100]: ack. done.

File: sdks/go/pkg/beam/io/textio/textio.go (hunk 2 above)
Comment (ceocoder) [reply sdks/go/pkg/beam/io/textio/textio.go:99]: ack. done.

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by lostluck** [comment]
Thank you for contributing to Beam! Welcome to the community! :)

**Pull request merged**